confighelper/beaconbroadcaster.py

```diff
@@ -0,0 +1,129 @@
+        uid: The UID of the calling program
+        gid: The GID of the calling program
+        """
+
+        self.logger = logging.getLogger(__name__)
+
+        self.logger.debug("Initializing broadcaster for beacon %s from program %s.",
+                          beacon_name, program_name)
+
+        self.beacon_path = os.path.join(BEACON_PATH, program_name, beacon_name)
+        self.program_name = program_name
+        self.beacon_name = beacon_name
+
+        tmpfs_path = os.path.join(BEACON_PATH, program_name)
+        tmpfs.mount_tmpfs(tmpfs_path, TMPFS_SIZE)
+
+        self._create_directory(self.beacon_path)
+        self._set_file_permissions(self.beacon_path, GROUP_RW_MODE, uid, gid)
+
+        self.logger.info("Broadcaster for eacon %s from program %s initialized.",
+                         beacon_name, program_name)
+
+    def send(self):
+        """ Place a new file in the beacon directory. Will raise an exception if it fails."""
+        self.logger.info(
+            "Sending beacon %s for program %s.", self.beacon_name, self.program_name)
+        now = datetime.datetime.now().isoformat()
+        random_number = os.urandom(16).encode('hex')
+
+        beacon_filename = '%s---%s' % (now, random_number)
+        final_path = os.path.join(self.beacon_path, beacon_filename)
+
+        try:
+            previous_beacons = os.listdir(self.beacon_path)
+            open(final_path, 'a').close()
+
+            for beacon_file in previous_beacons:
+                os.remove(os.path.join(self.beacon_path, beacon_file))
+
+        except Exception as exception:
+            message = 'Could not create beacon file for beacon %s, program %s. %s:%s' % (
+                self.beacon_name, self.program_name, type(exception).__name__,
+                str(exception))
+            self.logger.critical(message)
+            raise BeaconBroadcasterBroadcastException(message)
+
+    def _set_file_permissions(self, path, mode, uid, gid):
+        """ Sets permissions for a file. Will raise an exception if it fails.
+
+        path: The file to modify
+        mode: The mode to set
+        uid: The owner to set
+        gid: The group to set
+        """
+        try:
+            os.chown(path, uid, gid)
+            # Set permissions to xrw-------
+            os.chmod(path, mode)
+
+        except Exception as exception:
+            self.logger.critical('Could not set permissions for file %s for beacon %s for'
+                                 'program %s. %s: %s.', path, self.beacon_name,
+                                 self.program_name, type(exception).__name__, str(exception))
+            raise BeaconBroadcasterInitException
+
+    def _create_directory(self, path):
+        """ Create a directory while handling errors. Will raise an exception if it fails.
+
+        path: The directory to create
+        """
+        path = os.path.abspath(path)
+        if not os.path.isdir(path):
```

Comment:
What if it is a file?

Reply:
This has been replaced with your create_directory function from gpgmailer.